src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -277,35 +404,129 @@ public override void ClearAssetCache()
+                    ids.Contains(idPair.Item2))
                 {
-                    if (AssetDatabase.TryGetGUIDAndLocalFileIdentifier(asset, out string guid, out long localid))
+                    ids.Remove(idPair.Item2);
+                    if (ids.Count == 0)
                     {
-                        oldAssets.Add(asset, new TAssetEntry
-                        {
-                            AssetId = new AssetId(new Guid(guid), localid),
-                            Asset = asset
-                        });
+                        string assetName = GetAssetCachesContentPath($"{GetValidAssetName(idPair.Item1)}_{this.GetType().Name}", assetContentExtension);
+                        AssetDatabase.DeleteAsset(assetName);
                     }
-                    else
+                }
+            }
+
+            lookupByName = oldAssets;
+            var tempAssets = new List<NameEntry>();
+            foreach (var item in lookupByName)
+            {
+                tempAssets.Add(new NameEntry(item.Key, item.Value.OrderBy(a => a.Guid).ThenBy(a => a.FileIdentifier).ToArray()));
+            }
+            assets = tempAssets.OrderBy(a => a.Name).ToArray();
+            EditorUtility.SetDirty(this);
+#endif
+        }
+
+        public virtual void SaveAssets()
+        {
+#if UNITY_EDITOR
+            EnsureAssetDirectoryExists();
+            if (assets == null ||
+                assets.Length == 0)
+            {
+                Debug.Log("Assets was null or empty, rerunning asset cache update");
+                UpdateAssetCache();
+            }
+
+            if (assets == null ||
+                assets.Length == 0)
+            {
+                Debug.Log("No assets were found.");
+                return;
+            }
+
+            foreach (var nameEntry in assets)
+            {
+                string assetName = GetAssetCachesContentPath($"{GetValidAssetName(nameEntry.Name)}_{this.GetType().Name}", assetContentExtension);
+                AssetCacheContent content = ScriptableObject.CreateInstance<AssetCacheContent>();
+                List<AssetCacheEntry> assetCacheEntries = new List<AssetCacheEntry>();
+                foreach (var assetId in nameEntry.Ids)
+                {
+                    if (!lookupByAssetId.TryGetValue(assetId, out var entry))
                     {
-                        Debug.LogError($"Unable to obtain id for Asset: {asset.ToString()}");
+                        Debug.LogError($"AssetId did not have a registered Asset: {assetId}");
+                        continue;
                     }
+
+                    assetCacheEntries.Add(entry);
                 }
+
+                content.AssetCacheEntries = assetCacheEntries.ToArray();
+                Debug.Log($"Creating asset: {content} {assetName}");
+                AssetDatabase.CreateAsset(content, assetName);
+            }
+#endif
+        }
+
+        string GetValidAssetName(string name)

```

Comment:
>GetValidAssetName [](start = 15, length = 17)

Is it possible for this function to create naming collisions? i.e. because we're removing certain illegal characters (which may have made them unique before), they now become non-unique? If this happens, how will this manifest?

Reply:
We can already end up with non-unique name collisions. We currently load all assets with the same name that are of the same type at the same time. Once loaded, they will have different asset guids/file identifiers. 

---
In reply to: [355049066](https://github.com/microsoft/MixedReality-SpectatorView/pull/295#discussion_r355049066) [](ancestors = 355049066)